Allow subdomains, block or allow subdirectories

penge opened the pull request:
Closes https://github.com/penge/block-site/issues/12
Closes https://github.com/penge/block-site/issues/9
Closes https://github.com/penge/block-site/issues/5

Inspired by https://github.com/penge/block-site/pull/11 and https://github.com/penge/block-site/pull/7

Huge thanks to @clairdl as I revisited his suggestions on making simpler descriptions, and I completely agree. Also, I'd like to thank to @lumoe for the initiation that inspired me to continue on that and solve currently open issues.

penge commented:
I'd like to invite @lumoe and @clairdl for a review.

penge commented:
Should I call websites we block "time-consuming websites" or "distracting websites"?

penge commented on README.md:
```diff
@@ -1,21 +1,18 @@
 # Block Site
 
-**Block Site** (formerly known as Stop Social Media) is a simple Chrome extension that improves your productivity by blocking the access to time-consuming websites (like during the working hours) as you specify in **Options**. **Block Site** is by default disabled and doesn't block any website until you say so in **Options**.
+**Block Site** (formerly known as Stop Social Media) is a simple Chrome extension that improves your productivity by blocking access to distracting websites as you specify.
```
Should I call websites we block "time-consuming websites" or "distracting websites"?

lumoe replied:
I would suggest "distracting websites". For me, the goal of this extension is not to prevent me from spending a lot of time on the website, but from going there in the first place.

penge replied:
Let's call it "distracting websites" then. I agree.

I'll update manifest file (as part of this PR) and repository description (once the PR is merged).

penge commented on background.js:
```diff
@@ -1,31 +1,66 @@
+// "https://www.youtube.com/feed/explore" => "youtube.com/feed/explore"
+// "https://music.youtube.com/" => "music.youtube.com"
+// "https://music.youtube.com/explore" => "music.youtube.com/explore"
+const getDomainUrl = (url) => {
+  let domainUrl = url.replace(/^http(s?):\/\//, "").replace(/^www\./, "");
+  domainUrl = domainUrl.endsWith("/") ? domainUrl.slice(0, -1) : domainUrl;
+  return domainUrl;
+};
```
`www.` is not important (and is stripped), other subdomains are important, like `music.`.

penge commented on background.js:
```diff
@@ -1,31 +1,66 @@
+  const rules = [
+    ...allowList.map((path) => ({ path, type: "allow" })),
+    ...blockList.map((path) => ({ path, type: "block" })),
+  ].sort((a, b) => b.path.length - a.path.length);
```
I am sorting the rules by their length, as their length defines their specificity.
**Example:** `music.youtube.com` is more specific than `youtube.com` and we might want to allow `music.youtube.com` while still block `youtube.com`, so we should check `music.youtube.com` rule first (otherwise `youtube.com` rule would block `music.youtube.com`).

lumoe replied:
Sorting the rules by length is an elegant solution to order by specificity!

penge replied:
Thanks!

penge commented on background.js:
```diff
@@ -1,31 +1,66 @@
-      chrome.tabs.remove(tabId);
+    const rules = getRules(blocked);
+    const foundRule = rules.find((rule) => domainUrl.startsWith(rule.path) || domainUrl.endsWith(rule.path));
+    if (!foundRule || foundRule.type === "allow") {
```
The most specific matching rule is evaluated.

lumoe commented:
Thank you for your effort, also your solution to support filtering by `pathname` is great. Tested the `subdirectories` branch with different filter settings on my machine (Ubuntu 20.04 and Chrome 97).

Two things I would suggest:

1. There should be an example in the README showing that it is possible to have rules with pathname in it
2. Run the URL normalizing function `getDomainUrl` also on the rules. Having the rule `https://youtube.com/` currently does not block the site. After running the function it would become  `youtube.com` which would then result in a matching rule.

lumoe commented on README.md:
```diff
@@ -1,21 +1,18 @@
 youtube.com
+!music.youtube.com
 twitter.com
 reddit.com
```
Adding an example for `pathname` would help. Using subreddits on `reddit.com` for this would be a great example. 
e.g. `!reddit.com/r/MachineLearning/`

penge replied:
Great example! Let's have it. If I may suggest, let's modify the example just a slight bit, **without a trailing slash,** as:

```
!reddit.com/r/MachineLearning
```

Which now explains two things. Not only it shows it **allows** `reddit.com/r/MachineLearning`, but also it shows that trailing slash is **not required** to make this work (as reddit redirects it to `reddit.com/r/MachineLearning/`).

penge commented:
@lumoe

Thank you for your great suggestions! I have updated the rules to support fully qualified URLs as well, example:

```
https://youtube.com/
!https://music.youtube.com/
```

This may be more natural to some, as simple copy & paste URL will now work. Should we document it and provide more examples?

On one side, I don't want to confuse the user with too many examples. On other side, it is a worthy feature to mention. What we could do, is to separate examples with an empty line, and put same examples with fully qualified URLs below. What do you think?

I have added your `!reddit.com/r/MachineLearning` which I think is great. There's one more example we need. That is, blocking subpath (subdirectory) only. Any suggestions? @lumoe @clairdl

penge commented:
Going to ship it. Summarizing the new features here. You can now:

1. Exclude subdomain from blocking (https://github.com/penge/block-site/issues/12), example:

```
youtube.com
!music.youtube.com
```

The above example would block `youtube.com`, but allow `music.youtube.com`.

2. Block subdirectories only (https://github.com/penge/block-site/issues/9), example:

```
youtube.com/user/BBC
```

The above example would block `youtube.com/user/BBC` only, and so allow `youtube.com`.

3. Allow subdirectories only (https://github.com/penge/block-site/issues/5), example:

```
youtube.com
!youtube.com/results?search_query
```

The above example would block `youtube.com`, but allow `youtube.com/results?search_query`.

<br>

**Note:** Fully qualified URLs will work too (including `https` or `https`, `www`). Trailing slash is optional.
